components/layout_2020/positioned.rs

```diff
@@ -482,55 +481,125 @@ impl HoistedAbsolutelyPositionedBox {
-                        inline_size,
-                        block_size: block_axis.size,
-                        style: &non_replaced.style,
-                    };
-                    // https://drafts.csswg.org/css-writing-modes/#orthogonal-flows
-                    assert_eq!(
-                        containing_block.style.writing_mode,
-                        containing_block_for_children.style.writing_mode,
-                        "Mixed writing modes are not supported yet"
-                    );
-                    let dummy_tree_rank = 0;
-                    let independent_layout = non_replaced.layout(
-                        layout_context,
-                        &mut positioning_context,
-                        &containing_block_for_children,
-                        dummy_tree_rank,
-                    );
-
-                    content_size = Vec2 {
-                        inline: inline_size,
-                        block: block_axis
-                            .size
-                            .auto_is(|| independent_layout.content_block_size),
+                    // If the tentative used inline size is greater than ‘max-inline-size’,
+                    // recalculate the inline size and margins with ‘max-inline-size’ as the
+                    // computed ‘inline-size’. We can assume the new inline size won’t be ‘auto’,
+                    // because a non-‘auto’ computed ‘inline-size’ always becomes the used value.
+                    if let Some(max) = max_size.inline {
+                        if inline_size > max {
+                            inline_axis = solve_inline_axis(LengthOrAuto::LengthPercentage(max));
+                            inline_size = inline_axis.size.auto_is(|| unreachable!());
+                        }
+                    }
+
+                    // If the tentative used inline size is less than ‘min-inline-size’,
+                    // recalculate the inline size and margins with ‘min-inline-size’ as the
+                    // computed ‘inline-size’. We can assume the new inline size won’t be ‘auto’,
+                    // because a non-‘auto’ computed ‘inline-size’ always becomes the used value.
+                    if inline_size < min_size.inline {
+                        inline_axis =
+                            solve_inline_axis(LengthOrAuto::LengthPercentage(min_size.inline));
+                        inline_size = inline_axis.size.auto_is(|| unreachable!());
+                    }
+
+                    struct Result {
+                        content_size: Vec2<CSSPixelLength>,
+                        fragments: Vec<Fragment>,
+                    }
+
+                    let mut try_layout = |size| {
+                        let containing_block_for_children = ContainingBlock {
+                            inline_size,
+                            block_size: size,
+                            style: &non_replaced.style,
+                        };
```

Comment:
do we really need to relayout the children with the new block size, or can we just recalculate the block size and margins, but keep the old fragments and containing block?

Reply:
d77abd6bbe2161170dbfa40285012717745209d4 avoids the relayout with a new containing block size, which simplifies the code and doesn’t affect any tests in /css/CSS2/positioning/, but i feel like it might be incorrect because the containing block size affects percentages, no?

Reply:
it looks like the answer is yes, though i can’t quite say exactly where the CSS2 spec says so. this test page yields a green square in firefox and d6c884574e4905e711bcd80da37b55e19915f99d, but a black square in layout 2013 and d6c884574e4905e711bcd80da37b55e19915f99d:

```html
<!doctype html><meta charset="utf-8">
<p>There should be a green square below.
<div style="position: absolute; width: 0; height: 0; min-width: 1em; min-height: 1em; background: black;">
    <div style="width: 100%; height: 100%; background: green;"></div>
</div>
```

https://bucket.daz.cat/work/igalia/servo/2.html

Reply:
47180bb06493e7a1396be8f048491335d288f4a1 fixes this and adds some tests :)